File: solutions/code-carl/ch08_greedy/T12_LC0452_FindMinArrowShots.java
/**
 * Practice template for:
 * - LeetCode 452. Minimum Number of Arrows to Burst Balloons
 *   https://leetcode.cn/problems/minimum-number-of-arrows-to-burst-balloons/
 */
import java.util.Arrays;
import java.util.Map;

/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2025/4/7 10:11
 * @comment
 */
public class T12_LC0452_FindMinArrowShots {
    public int findMinArrowShots(int[][] points) {
        Arrays.sort(points, (a, b) -> Integer.compare(a[0], b[0]));
        int rightBorder = points[0][1];
        int count = 1;

        for (int i = 1; i < points.length; i++) {
            if (points[i][0] <= rightBorder) {
                rightBorder = Math.min(rightBorder, points[i][1]);
            } else {
                rightBorder = points[i][1];
                count++;
            }
        }
        return count;
    }
}

File: templates/code-carl/ch08_greedy/t12_lc0452_find_min_arrow_shots.py
"""LeetCode 452. Minimum Number of Arrows to Burst Balloons
https://leetcode.cn/problems/minimum-number-of-arrows-to-burst-balloons/
"""

from typing import List


class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        pass

if __name__ == "__main__":
    solution = Solution()
    assert solution.findMinArrowShots([[10, 16], [2, 8], [1, 6], [7, 12]]) == 2
    assert solution.findMinArrowShots([[1, 2], [3, 4], [5, 6], [7, 8]]) == 4
    assert solution.findMinArrowShots([[1, 2], [2, 3], [3, 4], [4, 5]]) == 2
    print("All tests passed.")
